src/libxmljs.cc

```diff
@@ -50,66 +88,66 @@ void xmlMemFreeWrap(void* p)
 #if (NODE_MODULE_VERSION > 0x000B)
     if (v8::Isolate::GetCurrent() == 0)
     {
+        assert(diff <= 0);
         return;
     }
 #endif
     if (v8::V8::IsDead())
     {
+        assert(diff <= 0);
         return;
     }
-
-    const int diff = xmlMemUsed() - xml_memory_used;
-    xml_memory_used += diff;
     NanAdjustExternalMemory(diff);
 }
 
-// wrapper for xmlMemRealloc to update v8's knowledge of memory used
-void* xmlMemReallocWrap(void* ptr, size_t size)
+void* memMalloc(size_t size)
 {
-    void* res = xmlMemRealloc(ptr, size);
-
-    // if realloc fails, no need to update v8 memory state
-    if (!res)
-    {
-        return res;
-    }
-
-    const int diff = xmlMemUsed() - xml_memory_used;
-    xml_memory_used += diff;
-    NanAdjustExternalMemory(diff);
-    return res;
+    size_t totalSize = size + HDR_SIZE;
+    memHdr* mem = static_cast<memHdr*>(malloc(totalSize));
+    if (!mem) return NULL;
+    mem->size = size;
+    adjustMem(totalSize);
+    return hdr2client(mem);
 }
 
-// wrapper for xmlMemoryStrdupWrap to update v8's knowledge of memory used
-char* xmlMemoryStrdupWrap(const char* str)
+void memFree(void* p)
 {
-    char* res = xmlMemoryStrdup(str);
+    memHdr* mem = client2hdr(p);
+    ssize_t totalSize = mem->size + HDR_SIZE;
+    adjustMem(-totalSize);
+    free(mem);
+}
 
-    // if strdup fails, no need to update v8 memory state
-    if (!res)
-    {
-        return res;
-    }
+void* memRealloc(void* ptr, size_t size)
+{
+    memHdr* mem1 = client2hdr(ptr);
+    ssize_t oldSize = mem1->size;
+    memHdr* mem2 = static_cast<memHdr*>(realloc(mem1, size + HDR_SIZE));
+    if (!mem2) return NULL;
+    mem2->size = size;
+    adjustMem(ssize_t(size) - oldSize);
+    return hdr2client(mem2);
+}
 
-    const int diff = xmlMemUsed() - xml_memory_used;
-    xml_memory_used += diff;
-    NanAdjustExternalMemory(diff);
+char* memStrdup(const char* str)
+{
+    size_t size = strlen(str) + 1;
+    char* res = static_cast<char*>(memMalloc(size));
+    if (res) strcpy(res, str);
```

Comment:
This should be a memcpy, given that the string length has already been determined.